File: 2021/day15.cpp
// https://adventofcode.com/2021/day/15

#include "../common/common.h"
#include "../common/rust.h"

#include <iostream>
#include <print>
#include <ranges>
#include <string>

using Input = aoc::char_grid<>;

auto parse(String const& filename) -> Input {
  return aoc::read_char_grid(filename);
}

fn solve_case1(Input const& grid) -> int {
  let start = point{0, 0};
  let end = point{static_cast<int>(grid.num_columns()) - 1,
                  static_cast<int>(grid.num_rows()) - 1};
  let distances = aoc::shortest_distances_dijkstra(
      start,
      [&](point current) {
        return grid.basic_neighbor_positions(current) |
               std::views::transform([&](point neighbor) {
                 return aoc::dijkstra_neighbor_t{
                     neighbor, grid.at(neighbor.y, neighbor.x) - '0'};
               });
      },
      end);
  return distances.find(end)->second;
}

fn virtual_cost(Input const& grid, point pos) -> int {
  let base = grid.at(pos.y % grid.num_rows(), pos.x % grid.num_columns()) - '0';
  let offset = pos.y / static_cast<int>(grid.num_rows()) +
               pos.x / static_cast<int>(grid.num_columns());
  return (base + offset - 1) % 9 + 1;
}

fn virtual_neighbors(point pos, int rows, int cols)
    -> aoc::static_vector<point, 4> {
  auto neighbors = aoc::static_vector<point, 4>{};
  if (pos.y > 0) {
    neighbors.push_back({pos.x, pos.y - 1});
  }
  if (pos.x > 0) {
    neighbors.push_back({pos.x - 1, pos.y});
  }
  if (pos.y + 1 < rows) {
    neighbors.push_back({pos.x, pos.y + 1});
  }
  if (pos.x + 1 < cols) {
    neighbors.push_back({pos.x + 1, pos.y});
  }
  return neighbors;
}

fn solve_case2(Input const& grid) -> int {
  let rows = static_cast<int>(grid.num_rows()) * 5;
  let cols = static_cast<int>(grid.num_columns()) * 5;
  let start = point{0, 0};
  let end = point{cols - 1, rows - 1};
  let distances = aoc::shortest_distances_dijkstra(
      start,
      [&](point current) {
        return virtual_neighbors(current, rows, cols) |
               std::views::transform([&](point neighbor) {
                 return aoc::dijkstra_neighbor_t{neighbor,
                                                 virtual_cost(grid, neighbor)};
               });
      },
      end);
  return distances.find(end)->second;
}

int main() {
  std::println("Part 1");
  let example = parse("day15.example");
  AOC_EXPECT_RESULT(40, solve_case1(example));
  let input = parse("day15.input");
  AOC_EXPECT_RESULT(592, solve_case1(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(315, solve_case2(example));
  AOC_EXPECT_RESULT(2897, solve_case2(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2021/src/day15.rs
use aoc::dijkstra::DijkstraState;
use aoc::grid::Grid;
use aoc::grid::Upos;
use aoc::iter::CollectArrayVec;
use arrayvec::ArrayVec;

type Input = Grid<u32>;

fn parse(filename: &str) -> Input {
    aoc::grid::from_file(filename)
}

fn solve_case1(grid: &Input) -> u32 {
    let start = Upos { y: 0, x: 0 };
    let end = Upos {
        y: grid.num_rows - 1,
        x: grid.num_columns - 1,
    };
    let distances = aoc::dijkstra::shortest_distances(
        &start,
        |current: &Upos| *current == end,
        |current: &Upos| {
            grid.basic_neighbor_positions(*current)
                .iter()
                .map(|&pos| DijkstraState {
                    distance: *grid.get(pos.y, pos.x),
                    data: pos,
                })
                .collect_array_vec::<4>()
        },
    );
    return distances[&end];
}

fn virtual_neighbors(pos: Upos, rows: usize, cols: usize) -> ArrayVec<Upos, 4> {
    let mut neighbors = ArrayVec::new();
    if pos.y > 0 {
        neighbors.push(Upos {
            y: pos.y - 1,
            x: pos.x,
        });
    }
    if pos.x > 0 {
        neighbors.push(Upos {
            y: pos.y,
            x: pos.x - 1,
        });
    }
    if pos.y + 1 < rows {
        neighbors.push(Upos {
            y: pos.y + 1,
            x: pos.x,
        });
    }
    if pos.x + 1 < cols {
        neighbors.push(Upos {
            y: pos.y,
            x: pos.x + 1,
        });
    }
    neighbors
}

fn virtual_cost(grid: &Input, pos: Upos) -> u32 {
    let base = *grid.get(pos.y % grid.num_rows, pos.x % grid.num_columns);
    let offset = (pos.y / grid.num_rows + pos.x / grid.num_columns) as u32;
    (base + offset - 1) % 9 + 1
}

fn solve_case2(grid: &Input) -> u32 {
    let rows = grid.num_rows * 5;
    let cols = grid.num_columns * 5;
    let start = Upos { y: 0, x: 0 };
    let end = Upos {
        y: rows - 1,
        x: cols - 1,
    };
    let distances = aoc::dijkstra::shortest_distances(
        &start,
        |current: &Upos| *current == end,
        |current: &Upos| {
            virtual_neighbors(*current, rows, cols)
                .iter()
                .map(|&pos| DijkstraState {
                    distance: virtual_cost(grid, pos),
                    data: pos,
                })
                .collect_array_vec::<4>()
        },
    );
    return distances[&end];
}

fn main() {
    println!("Part 1");
    let example = parse("day15.example");
    assert_eq!(40, solve_case1(&example));
    let input = parse("day15.input");
    assert_eq!(592, solve_case1(&input));

    println!("Part 2");
    assert_eq!(315, solve_case2(&example));
    assert_eq!(2897, solve_case2(&input));
}
